src/gflip/gflip_engine.cpp

```diff
@@ -358,10 +443,134 @@ void gflip_engine::run_evaluation(int dtype)
 
 // ---------------------------------------------------------
 
+void gflip_engine::update_tfidf()
+{
+  {
+    ScopedTimer scopedTimer("Adding docs");
+    for (int i=0; i<laserscan_bow.size(); ++i)
+    {
+      add_doc_stats(i);
+    }
+  }
+
+  {
+    ScopedTimer scopedTimer("Updating tf_idf");
+    compute_idfs();
+    compute_tf_idfs();
+  }
+
+  {
+    ScopedTimer scopedTimer(" tf_idf");
+    for(uint i=0;i<laserscan_bow.size();i++)
+      laserscan_bow[i].norm_wgv = norm_gfp(laserscan_bow[i]);
+  }
+
+}
+
+// ---------------------------------------------------------
+
+
+void gflip_engine::update_tfidf(int nscans)
+{
+  {
+    ScopedTimer scopedTimer("Adding scans");
+    for (int i=laserscan_bow.size()- nscans; i<laserscan_bow.size(); ++i)
+    {
+      add_doc_stats(i);
+    }
+  }
+  {
+    ScopedTimer scopedTimer("Updating tf_idf");
+    compute_idfs();
+    compute_tf_idfs();
+  }
+
+  {
+    ScopedTimer scopedTimer("Normalizing tf_idf");
+    for(uint i=0;i<laserscan_bow.size();i++)
+      laserscan_bow[i].norm_wgv = norm_gfp(laserscan_bow[i]);
+  }
+}
+
+
+
+// ---------------------------------------------------------
+
+void gflip_engine::add_doc_stats(int doc_id)
+{
+
+    std::map<int, std::vector<int> >::const_iterator it = laserscan_bow[doc_id].histogram.begin();
+    for(; it != laserscan_bow[doc_id].histogram.end(); ++it)
+    {
+      int word_id = it->first;
+      tf_idf[word_id].term_count_unnormalized.push_back(it->second.size());
+      tf_idf[word_id].num_words.push_back(laserscan_bow[doc_id].w.size());
```

Comment:
`laserscan_bow[doc_id].w.size()` doesn't change inside the loop, you can use a local variable.